src/15-budget-intent/index.ts

```diff
@@ -0,0 +1,167 @@
+
+const chalkFn: Record<string, (...args: string[]) => void> = {
+  turbo: chalk.green,
+  fourO: chalk.green.dim,
+  haiku: chalk.blue,
+  opus: chalk.blue.dim,
+  sonnet: chalk.cyan,
+  sonnet35: chalk.cyan.dim,
+}
+
+const system = dedent`
+  Your task is to analyze a plaintext budget statement
+  and produce a structured object to describe it.
+
+  If a budget expresses a range, set min and max accordingly.
+
+  If a budget expresses a single amount, then set min to 0 and max to the amount.
+
+  If a budget cannot be inferred, set both min and max to -1.
+`
+
+const prompt = (statement: string) => dedent`
+  Analyze the following budget statement and produce a structured object to describe it.
+
+  BUDGET STATEMENT: ${statement}
+`
+const schema = z.object({
+  budget: z
+    .object({
+      currency: z
+        .string()
+        .default("USD")
+        .describe("ISO currency code. Assume USD if not otherwise specified."),
+      min: z
+        .number()
+        .default(0)
+        .describe(
+          "A minimum amount. Default to 0 if not otherwise specified. Set this to a non-zero number ONLY IF the user has explicitly specified a minimum."
+        ),
+      max: z.number().describe("A maximum amount."),
+    })
+    .describe("The user's budget. If unspecified, set to the range [-1, -1]."),
+  numberOfArtworks: z
+    .number()
+    .default(1)
+    .describe(
+      "The number of artworks the user wishes to buy. Default 1 to unless otherwise specified."
+    ),
+  timeSpan: z
+    .object({
+      number: z.number().optional().describe("How many of the time units"),
+      unit: z
+        .enum(["day", "week", "month", "year"])
+        .optional()
+        .describe("The unit of time"),
+    })
+    .optional()
+    .describe(
+      "The time span over which the user wishes to buy. Omit if unspecified."
+    ),
+})
+
+const doLLM = async (modelName: string, example: string) => {
+  const start = Date.now()
+
+  const intent = await generateObject({
+    model: MODELS[modelName],
+    temperature: 0,
+    system,
+    prompt: prompt(example),
+    schema,
+  })
+
+  const elapsedMS = Date.now() - start
+
+  const exampleNum = EXAMPLES.indexOf(example)
+  const modelNum = Object.keys(MODELS).indexOf(modelName)
+
+  console.log(
+    chalkFn[modelName](
+      new Date().toISOString(),
+      modelName,
+      example.slice(0, 40)
+    )
+  )
+
+  return {
+    exampleNum,
+    modelNum,
+    example,
+    modelName,
+    elapsedMS,
+    intent: JSON.stringify(intent.object),
+  }
+}
+
+async function main() {
+  const file = fs.openSync("budgetIntents.ndjson", "w")
+
+  /* Each example in series, each model in parallel */
+  const exampleThunks = EXAMPLES.map((example) => () => {
+    const modelThunks = _.keys(MODELS).map(
+      (modelName) => () => doLLM(modelName, example)
+    )
+
+    return pAll(modelThunks)
+  })
+
+  const exampleResults = await pSeries(exampleThunks)
+  const flatResults = _.flatten(exampleResults)
```

Comment:
Some minor trickiness here re: how to fire off all these requests without getting rate limited — which is what happens when all 30 get fired off immediately via async await.

Instead I wanted to do these in a series of batches.

Had to work my way through the utilities at https://github.com/sindresorhus/promise-fun in order to figure out the best way to do this, but basically I use `pSeries` and `pAll` to :

- fire off the first 5 model requests in parallel for example 1. once all those are received…
- fire off the next 5 model request in parallel for example 2. once all those are received…
- fire off the next 5 model request in parallel for example 3. once all those are received…
- and so on…